{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metageomics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The human microbiome is involved in many essential functions, such as food digestion and modulation if immune system. Many diseases have been associated with microbiota disbiosis among them, psoriasis and chorinc skin inflammation reguarding skin flora alteration, and colorectal cancers and  inflammatory bowel deseases, in association with colonic microbiota disbiosis. High troughtput DNA sequencing tecniques allowed to study the entire genome of a microbiotic population and also quantify its aboundance. \n",
    "\n",
    "The main steps of a microbila study are:\n",
    "<ol>\n",
    "     <li>Mirobial DNA extraction and sequencing, using two approches, amplicon sequencing (16S) or shotgun sequencing</li>\n",
    "    <li>Bioinformatic sequence processing</li>\n",
    "    <li>Statistical analysis</li>\n",
    "</ol>\n",
    "Amplicon sequencing relies on, after a PCR amplification, sequencing a phylogenetic marker gene, which is 16S ribosomal RNA gene, that encodes the small ribosomal subunit for the archea and bacteria.\n",
    "The 16S gene contains both high conserved areas and hypervariable sites, denoted as V1-V9.\n",
    "The conserved regions can be targeted with PCR primers, while hypervariable sites are specific for each microbial species(or OTU), which allows to dicriminate different microbes.\n",
    "V1-V3 or V4 are usually used as target in PCR amplification, from which thousands to millions of DNA copies are generated. PCR's amplicons are than sequenced by hightroughput tecniques, and multiples nucleotide sequenced called reads are obtained. \n",
    "\n",
    "The most used bioinformatic pipelines in amplicon sequencing are: mothur and <b>QIIME</b>. \n",
    "Bioinformatic processing consist in five steps:\n",
    "\n",
    "<ol>\n",
    "    <li>Preprocessing and quality control filtering</li>\n",
    "<li>OTU(operation taxonomic unit) binning </li>\n",
    "<li>Taxonomy assignment </li>\n",
    "<li>Construction of abundance table </li>\n",
    "<li>Phylogenetic analysis </li>\n",
    "</ol>\n",
    "\n",
    "Preprocessing and quality control filtering consists in assign sequences to samples, using the barcode contained in each sequence (<b>Demultiplexing</b>), then sequences are quality filtered to remove too short sequences, too many ambiguos base pairs and chimeras. \n",
    "\n",
    "OTU binning is the process of clustering similar DNA sequences into OTUs (groups of DNA sequnces with at least 97% similarity). The sequences assigned to OTUs are representative sequences determined by the most common nucleotide founded in each position.\n",
    "\n",
    "Taxonomic assignment consists in comparing OTU consensus sequences to microbial 16S rRNA refrence database such as Green Genes, SILVA or RDP. This step provides the available annotation of each OTU to the different taxonomy level. Many OTUs are not annoted expecially for lowest taxonomic levels. \n",
    "\n",
    "Each entry of the OTU abundance table contains the number of reads observed for each OTU for each sample.\n",
    "OTU table can be very sparse, if there are too many zeros(many OTUs observed only in a few samples) could be useful to aggomerate OTUs at a broader taxonomic level. \n",
    "\n",
    "Phylogenetic trees can be used to obtein phylogeetic distances between samples.\n",
    "\n",
    "Beacause of the experimental process and the quality control filtering, microbiome data are very noisy and the total number of reads for sample is vary variable, which requires normalization of data before comparing the abundance between samples. The maximum number of sequences reads that the DNA sequencer can provide costraints the total number of counts for sample. It can bias the real abunances of each taxonomic unit. \n",
    "\n",
    "Statistical analysis starts with normalization and procedes with an exploratory part which consists in the analisys of diversity measures and their visualization thought multivariate techniques to visualize species abundances in a low dimentional space.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## QIIME2\n",
    "\n",
    "All data must be imported as a <b>QIIME2 artifact</b> to be used by a QIIME2 action (with the exception of some metadata). All amplicon/metagenome sequencing experiments begin, at some point or another, as raw sequence data. This is probably FASTQ data, containing DNA sequences and quality scores for each base.\n",
    "We must demultiplex these reads to determine which sample each read came from.\n",
    "Reads should then be denoised into amplicon sequence variants (ASVs) or clustered into operational taxonomic units (OTUs) to achieve two goals:\n",
    "\n",
    "<ol>\n",
    "    <li>reducing sequence errors</li>\n",
    "\n",
    "   <li> dereplicated sequences</li>\n",
    "</ol>\n",
    "\n",
    "The resulting feature table and representative sequences are key pieces of data. A feature table is essentially a matrix of samples x observations, i.e., the number of times each “feature” (OTUs, ASVs, etc) is observed in each sample in a data set.\n",
    "We can do many things with this feature table. Common analyses include:\n",
    "Taxonomic classification of sequences (a.k.a., “what species are present?”)\n",
    "Alpha and beta diversity analyses, or measures of diversity within and between samples, respectively (a.k.a., “how similar are my samples?”)\n",
    "Many diversity analyses rely on the phylogenetic similarity between individual features. If you are sequencing phylogenetic markers (e.g., 16S rRNA genes), you can align these sequences to assess the phylogenetic relationship between each of your features.\n",
    "Differential abundance measurements determine which features (OTUs, ASVs, taxa, etc) are significantly more/less abundant in different experimental groups.\n",
    "\n",
    "<img src=\"https://docs.qiime2.org/2019.10/_images/overview.png\">\n",
    "\n",
    "### Demultiplexing\n",
    "Most nextgen sequencing instruments have the capacity to analyze hundreds or even thousands of samples in a single lane/run; we do so by multiplexing these samples, which is just a fancy word for mixing a whole bunch of stuff together. How do we know which sample each read came from? This is typically done by appending a unique barcode (index or tag) sequence to one or both ends of each sequence. Detecting these barcode sequences and mapping them back to the samples they belong to allows us to demultiplex our sequences.\n",
    "You should know which barcode belongs to each sample and include this barcode information in your sample metadata file. Paired-end reads need to be joined at some point in the analysis, this happens automatically during denoising with q2-dada2.\n",
    "\n",
    "\n",
    "### Denoising and clustering\n",
    "\n",
    "The names for these steps are very descriptive:\n",
    "We denoise our sequences to remove and/or correct noisy reads.\n",
    "It dereplicates input sequences to reduce repetition and file size/memory requirements in downstream steps (it keeps count of each replicate).\n",
    "It clusters sequences to collapse similar sequences (e.g., those that are ≥ 97% similar to each other) into single replicate sequences. This process, also known as <b>OTU picking</b>, was once a common procedure, used to simultaneously dereplicate but also perform a sort of quick-and-dirty denoising procedure (to capture stochastic sequencing and PCR errors, which should be rare and similar to more abundant centroid sequences). Use denoising methods instead if you can.\n",
    "\n",
    "### Denoising (ASVs)\n",
    "\n",
    "The denoising methods currently available in QIIME 2 include DADA2 and Deblur. You can learn more about those methods by reading the original publications for each. Examples of DADA2 exist in the moving pictures tutorial and Fecal Microbiome Transplant study tutorial (for single-end data) and Atacama soils tutorial (for paired-end data). Examples of Deblur exist in the moving pictures tutorial (for single-end data) and read joining tutorial (for paired-end data). Note that deblur (and also vsearch dereplicate-sequences) should be preceded by basic quality-score-based filtering, but this is unnecessary for dada2. Both Deblur and DADA2 contain internal chimera checking methods and abundance filtering, so additional filtering should not be necessary following these methods.\n",
    "\n",
    "To put it simply, these methods filter out noisy sequences, correct errors in marginal sequences (in the case of DADA2), remove chimeric sequences, remove singletons, join denoised paired-end reads (in the case of DADA2), and then dereplicate those sequences. The products of these operations are called <b>ASVs</b>.\n",
    "\n",
    "\n",
    "### Clustering (OTU)\n",
    "Next we will discuss clustering methods. Dereplication (the simplest clustering method, effectively producing 100% OTUs, i.e., all unique sequences observed in the dataset) and it is the necessary starting point to all other clustering methods in QIIME2.\n",
    "<b>q2-vsearch</b> implements three different OTU clustering strategies: de novo, closed reference, and open reference. \n",
    "\n",
    "All should be preceded by basic quality-score-based filtering and followed by chimera filtering and aggressive OTU filtering (the treacherous trio, a.k.a. the Bokulich method).\n",
    "\n",
    "The features produced by clustering methods are known as operational taxonomic units (OTUs), which is Esperanto for suboptimal, imprecise rubbish. \n",
    "\n",
    "While OTU clustering approaches attempt to blur similar sequences into an abstracted consensus sequence, thus minimizing the influence of any sequencing errors within the pool of reads, the Amplicon Sequence Variant (ASV) approach attempts to go the opposite direction. The ASV approach will start by determining which exact sequences were read and how many times each exact sequence was read. These data will be combined with an error model for the sequencing run, enabling the comparison of similar reads to determine the probability that a given read at a given frequency is not due to sequencer error. This creates, in essence, a p-value for each exact sequence, where the null-hypothesis is equivalent to that exact sequence being a consequence of sequencing error.\n",
    "\n",
    "### The Feature Table\n",
    "\n",
    "The final products of all denoising and clustering methods/workflows are a FeatureTable(Frequency feature table) artifact and a FeatureData(representative sequences) artifact. These are two of the most important artifacts in an amplicon sequencing workflow, and are used for many downstream analyses, as discussed below. Indeed, feature tables are crucial to any QIIME2 analysis, as the central record of all observations per sample. Such an important artifact deserves its own powerful plugin, q2-feature-table.\n",
    "\n",
    "### Taxonomy classification and taxonomic analyses\n",
    "\n",
    "It can be done by comparing our query sequences (i.e., our features, be they ASVs or OTUs) to a reference database of sequences with known taxonomic composition. Simply finding the closest alignment is not really good enough — because other sequences that are equally close matches or nearly as close may have different taxonomic annotations. \n",
    "\n",
    "So we use taxonomy classifiers to determine the closest taxonomic affiliation with some degree of confidence or consensus (which may not be a species name if one cannot be predicted with certainty!), based on alignment, k-mer frequencies, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Microbiome Compositional Data \n",
    "\n",
    "Microbiome data is compositional because the information that abundance tables contain is relative. In a microbiome abundance table, the total number of counts per sample is highly variable and constrained by the maximum number of DNA reads that the sequencer can provide. This total count constraint induces strong dependencies among the abundances of the different taxa; an increase in the abundance of one taxon implies the decrease of the observed number of counts for some of the other taxa so that the total number of counts does not exceed the specified sequencing depth. Moreover, observed raw abundances and the total number of reads per sample are non-informative since they represent only a fraction or random sample of the original DNA content in the environment. These characteristics of microbiome abundance data clearly fall into the notion of compositional data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: permute\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "This is vegan 2.5-6\n",
      "\n",
      "\n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotly’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotly’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘phyloseq’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    distance\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    sampleNames\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    distance\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    distance\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plotly’:\n",
      "\n",
      "    arrange, mutate, rename, summarise\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('qiime2R')\n",
    "library('vegan')\n",
    "library('ggplot2')\n",
    "library('plotly')\n",
    "library('DESeq2')\n",
    "library('phyloseq')\n",
    "library('reshape2')\n",
    "library('plyr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Expected 7 pieces. Missing pieces filled with `NA` in 6755 rows [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].”\n"
     ]
    }
   ],
   "source": [
    "#setwd('/home/adelaide/adelaide/Dottorato/analisi/gastroDataAnalisi/')\n",
    "\n",
    "source(\"function.R\")\n",
    "\n",
    "#definizione dei path dei file di output di QIIME2\n",
    "Tab <- 'DADA2_output/dada2_gastrodata_table.qza'\n",
    "Map  <- 'Metadata/Mapping.txt'\n",
    "Tree <- 'Phylogenetic_tree/rooted_tree.qza'\n",
    "Tax <- \"Taxonomic_assignment/taxonomy_classification_data_sklearn.qza\"\n",
    "\n",
    "\n",
    "\n",
    "# crea l'oggetto phyloseq \n",
    "phy <- qza_to_phyloseq(Tab, Tree, Tax, Map, tmp='Temp')\n",
    "\n",
    "#\n",
    "phy <- subset_samples( phy, IBD==\"0\" | IBD==\"1\")#  | IBD==\"2\" )\n",
    "\n",
    "#corregge l'importazione della tassonomia silva\n",
    "phy <- importaTassonomiaSilva(phy)\n",
    "\n",
    "\n",
    "# trasforma IBD (e ID_original) in un factor, per i grafici\n",
    "metadata <- sample_data(phy)\n",
    "metadata$IBD <- as.factor(metadata$IBD)\n",
    "metadata$ID_original <- as.factor(metadata$ID_original)\n",
    "sample_data(phy) <- metadata\n",
    "\n",
    "#agglomerazione per...\n",
    "phyF <-tax_glom(phy,\"Family\")\n",
    "\n",
    "phyG <-tax_glom(phy,\"Genus\")\n",
    "\n",
    "# trasforma in dati percentuali\n",
    "phyF_percentuali <- transform_sample_counts(phyF, function(x) {x/sum(x)*100} )\n",
    "phyG_percentuali <- transform_sample_counts(phyG, function(x) {x/sum(x)*100} )\n",
    "# prende la tassonomia (trasformata in matrice, poi in dataframe (altrimenti non funzionava) ) per ricavare gli ID dei clostridi\n",
    "tassonomiaF <- as(tax_table(phyF_percentuali),\"matrix\")\n",
    "tassonomiaF <- as.data.frame(tassonomiaF)\n",
    "\n",
    "tassonomiaG <- as(tax_table(phyG_percentuali),\"matrix\")\n",
    "tassonomiaG <- as.data.frame(tassonomiaG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'BarcodeSequence'</li>\n",
       "\t<li>'LinkerPrimerSequence'</li>\n",
       "\t<li>'ID_F'</li>\n",
       "\t<li>'ID_original'</li>\n",
       "\t<li>'Data_prelievo'</li>\n",
       "\t<li>'Prelievo'</li>\n",
       "\t<li>'Nato'</li>\n",
       "\t<li>'Eta'</li>\n",
       "\t<li>'Genere'</li>\n",
       "\t<li>'Fumo'</li>\n",
       "\t<li>'Familiarità_IBD_'</li>\n",
       "\t<li>'IBD'</li>\n",
       "\t<li>'Montreal_RCU_localizzazione'</li>\n",
       "\t<li>'Montreal_MDC_Eta'</li>\n",
       "\t<li>'Montreal_MDC_localizzazione'</li>\n",
       "\t<li>'Montreal_MDC_Fenotipo'</li>\n",
       "\t<li>'Malattia_perianale'</li>\n",
       "\t<li>'Terapia_biologica_al_momento_del_prelievo'</li>\n",
       "\t<li>'Immunosoppressore_al_momento_del_prelievo'</li>\n",
       "\t<li>'Altra_terapia_al_momento_del_prelievo'</li>\n",
       "\t<li>'Pregressa_tp_biologica'</li>\n",
       "\t<li>'Pregressa_tp_con_immunosoppressori'</li>\n",
       "\t<li>'Interventi_chirurgici_per_malattia'</li>\n",
       "\t<li>'Attività'</li>\n",
       "\t<li>'Manifestazioni_extraintestinali'</li>\n",
       "\t<li>'MEI_attive'</li>\n",
       "\t<li>'Tipo_MEI'</li>\n",
       "\t<li>'Calprotectina_.250_.mg.kg.'</li>\n",
       "\t<li>'PCR_.mg.L.'</li>\n",
       "\t<li>'Emoglobina_.gr.dl.'</li>\n",
       "\t<li>'Globuli_bianchi_.mila.microL.'</li>\n",
       "\t<li>'Colesterolo_totale_.mg.dl.__'</li>\n",
       "\t<li>'Colesterolo_LDL_.mg.dl.'</li>\n",
       "\t<li>'Colesterolo_HDL_.mg.dl.'</li>\n",
       "\t<li>'Trigliceridi_.mg.dl.__'</li>\n",
       "\t<li>'Glicemia_.mg.dl.'</li>\n",
       "\t<li>'Steatosi_epatica'</li>\n",
       "\t<li>'Attività_endoscopica_.SES_CD._Mayo.'</li>\n",
       "\t<li>'Sicus'</li>\n",
       "\t<li>'Peso'</li>\n",
       "\t<li>'Altezza'</li>\n",
       "\t<li>'Parto'</li>\n",
       "\t<li>'Calo_ponderale'</li>\n",
       "\t<li>'Aumento_peso'</li>\n",
       "\t<li>'Yogurt'</li>\n",
       "\t<li>'Pane'</li>\n",
       "\t<li>'Pasta'</li>\n",
       "\t<li>'Latticini'</li>\n",
       "\t<li>'Frutta.Verdua'</li>\n",
       "\t<li>'Carne'</li>\n",
       "\t<li>'Pesce'</li>\n",
       "\t<li>'Cereali'</li>\n",
       "\t<li>'Legumi'</li>\n",
       "\t<li>'Caffè'</li>\n",
       "\t<li>'Alcool'</li>\n",
       "\t<li>'Pizza'</li>\n",
       "\t<li>'Fumo_al_momento_del_prelievo'</li>\n",
       "\t<li>'Esercizio_Fisico'</li>\n",
       "\t<li>'Comorbidità_degne_di_nota'</li>\n",
       "\t<li>'Comorbidità_prichiatriche'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BarcodeSequence'\n",
       "\\item 'LinkerPrimerSequence'\n",
       "\\item 'ID\\_F'\n",
       "\\item 'ID\\_original'\n",
       "\\item 'Data\\_prelievo'\n",
       "\\item 'Prelievo'\n",
       "\\item 'Nato'\n",
       "\\item 'Eta'\n",
       "\\item 'Genere'\n",
       "\\item 'Fumo'\n",
       "\\item 'Familiarità\\_IBD\\_'\n",
       "\\item 'IBD'\n",
       "\\item 'Montreal\\_RCU\\_localizzazione'\n",
       "\\item 'Montreal\\_MDC\\_Eta'\n",
       "\\item 'Montreal\\_MDC\\_localizzazione'\n",
       "\\item 'Montreal\\_MDC\\_Fenotipo'\n",
       "\\item 'Malattia\\_perianale'\n",
       "\\item 'Terapia\\_biologica\\_al\\_momento\\_del\\_prelievo'\n",
       "\\item 'Immunosoppressore\\_al\\_momento\\_del\\_prelievo'\n",
       "\\item 'Altra\\_terapia\\_al\\_momento\\_del\\_prelievo'\n",
       "\\item 'Pregressa\\_tp\\_biologica'\n",
       "\\item 'Pregressa\\_tp\\_con\\_immunosoppressori'\n",
       "\\item 'Interventi\\_chirurgici\\_per\\_malattia'\n",
       "\\item 'Attività'\n",
       "\\item 'Manifestazioni\\_extraintestinali'\n",
       "\\item 'MEI\\_attive'\n",
       "\\item 'Tipo\\_MEI'\n",
       "\\item 'Calprotectina\\_.250\\_.mg.kg.'\n",
       "\\item 'PCR\\_.mg.L.'\n",
       "\\item 'Emoglobina\\_.gr.dl.'\n",
       "\\item 'Globuli\\_bianchi\\_.mila.microL.'\n",
       "\\item 'Colesterolo\\_totale\\_.mg.dl.\\_\\_'\n",
       "\\item 'Colesterolo\\_LDL\\_.mg.dl.'\n",
       "\\item 'Colesterolo\\_HDL\\_.mg.dl.'\n",
       "\\item 'Trigliceridi\\_.mg.dl.\\_\\_'\n",
       "\\item 'Glicemia\\_.mg.dl.'\n",
       "\\item 'Steatosi\\_epatica'\n",
       "\\item 'Attività\\_endoscopica\\_.SES\\_CD.\\_Mayo.'\n",
       "\\item 'Sicus'\n",
       "\\item 'Peso'\n",
       "\\item 'Altezza'\n",
       "\\item 'Parto'\n",
       "\\item 'Calo\\_ponderale'\n",
       "\\item 'Aumento\\_peso'\n",
       "\\item 'Yogurt'\n",
       "\\item 'Pane'\n",
       "\\item 'Pasta'\n",
       "\\item 'Latticini'\n",
       "\\item 'Frutta.Verdua'\n",
       "\\item 'Carne'\n",
       "\\item 'Pesce'\n",
       "\\item 'Cereali'\n",
       "\\item 'Legumi'\n",
       "\\item 'Caffè'\n",
       "\\item 'Alcool'\n",
       "\\item 'Pizza'\n",
       "\\item 'Fumo\\_al\\_momento\\_del\\_prelievo'\n",
       "\\item 'Esercizio\\_Fisico'\n",
       "\\item 'Comorbidità\\_degne\\_di\\_nota'\n",
       "\\item 'Comorbidità\\_prichiatriche'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BarcodeSequence'\n",
       "2. 'LinkerPrimerSequence'\n",
       "3. 'ID_F'\n",
       "4. 'ID_original'\n",
       "5. 'Data_prelievo'\n",
       "6. 'Prelievo'\n",
       "7. 'Nato'\n",
       "8. 'Eta'\n",
       "9. 'Genere'\n",
       "10. 'Fumo'\n",
       "11. 'Familiarità_IBD_'\n",
       "12. 'IBD'\n",
       "13. 'Montreal_RCU_localizzazione'\n",
       "14. 'Montreal_MDC_Eta'\n",
       "15. 'Montreal_MDC_localizzazione'\n",
       "16. 'Montreal_MDC_Fenotipo'\n",
       "17. 'Malattia_perianale'\n",
       "18. 'Terapia_biologica_al_momento_del_prelievo'\n",
       "19. 'Immunosoppressore_al_momento_del_prelievo'\n",
       "20. 'Altra_terapia_al_momento_del_prelievo'\n",
       "21. 'Pregressa_tp_biologica'\n",
       "22. 'Pregressa_tp_con_immunosoppressori'\n",
       "23. 'Interventi_chirurgici_per_malattia'\n",
       "24. 'Attività'\n",
       "25. 'Manifestazioni_extraintestinali'\n",
       "26. 'MEI_attive'\n",
       "27. 'Tipo_MEI'\n",
       "28. 'Calprotectina_.250_.mg.kg.'\n",
       "29. 'PCR_.mg.L.'\n",
       "30. 'Emoglobina_.gr.dl.'\n",
       "31. 'Globuli_bianchi_.mila.microL.'\n",
       "32. 'Colesterolo_totale_.mg.dl.__'\n",
       "33. 'Colesterolo_LDL_.mg.dl.'\n",
       "34. 'Colesterolo_HDL_.mg.dl.'\n",
       "35. 'Trigliceridi_.mg.dl.__'\n",
       "36. 'Glicemia_.mg.dl.'\n",
       "37. 'Steatosi_epatica'\n",
       "38. 'Attività_endoscopica_.SES_CD._Mayo.'\n",
       "39. 'Sicus'\n",
       "40. 'Peso'\n",
       "41. 'Altezza'\n",
       "42. 'Parto'\n",
       "43. 'Calo_ponderale'\n",
       "44. 'Aumento_peso'\n",
       "45. 'Yogurt'\n",
       "46. 'Pane'\n",
       "47. 'Pasta'\n",
       "48. 'Latticini'\n",
       "49. 'Frutta.Verdua'\n",
       "50. 'Carne'\n",
       "51. 'Pesce'\n",
       "52. 'Cereali'\n",
       "53. 'Legumi'\n",
       "54. 'Caffè'\n",
       "55. 'Alcool'\n",
       "56. 'Pizza'\n",
       "57. 'Fumo_al_momento_del_prelievo'\n",
       "58. 'Esercizio_Fisico'\n",
       "59. 'Comorbidità_degne_di_nota'\n",
       "60. 'Comorbidità_prichiatriche'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"BarcodeSequence\"                          \n",
       " [2] \"LinkerPrimerSequence\"                     \n",
       " [3] \"ID_F\"                                     \n",
       " [4] \"ID_original\"                              \n",
       " [5] \"Data_prelievo\"                            \n",
       " [6] \"Prelievo\"                                 \n",
       " [7] \"Nato\"                                     \n",
       " [8] \"Eta\"                                      \n",
       " [9] \"Genere\"                                   \n",
       "[10] \"Fumo\"                                     \n",
       "[11] \"Familiarità_IBD_\"                         \n",
       "[12] \"IBD\"                                      \n",
       "[13] \"Montreal_RCU_localizzazione\"              \n",
       "[14] \"Montreal_MDC_Eta\"                         \n",
       "[15] \"Montreal_MDC_localizzazione\"              \n",
       "[16] \"Montreal_MDC_Fenotipo\"                    \n",
       "[17] \"Malattia_perianale\"                       \n",
       "[18] \"Terapia_biologica_al_momento_del_prelievo\"\n",
       "[19] \"Immunosoppressore_al_momento_del_prelievo\"\n",
       "[20] \"Altra_terapia_al_momento_del_prelievo\"    \n",
       "[21] \"Pregressa_tp_biologica\"                   \n",
       "[22] \"Pregressa_tp_con_immunosoppressori\"       \n",
       "[23] \"Interventi_chirurgici_per_malattia\"       \n",
       "[24] \"Attività\"                                 \n",
       "[25] \"Manifestazioni_extraintestinali\"          \n",
       "[26] \"MEI_attive\"                               \n",
       "[27] \"Tipo_MEI\"                                 \n",
       "[28] \"Calprotectina_.250_.mg.kg.\"               \n",
       "[29] \"PCR_.mg.L.\"                               \n",
       "[30] \"Emoglobina_.gr.dl.\"                       \n",
       "[31] \"Globuli_bianchi_.mila.microL.\"            \n",
       "[32] \"Colesterolo_totale_.mg.dl.__\"             \n",
       "[33] \"Colesterolo_LDL_.mg.dl.\"                  \n",
       "[34] \"Colesterolo_HDL_.mg.dl.\"                  \n",
       "[35] \"Trigliceridi_.mg.dl.__\"                   \n",
       "[36] \"Glicemia_.mg.dl.\"                         \n",
       "[37] \"Steatosi_epatica\"                         \n",
       "[38] \"Attività_endoscopica_.SES_CD._Mayo.\"      \n",
       "[39] \"Sicus\"                                    \n",
       "[40] \"Peso\"                                     \n",
       "[41] \"Altezza\"                                  \n",
       "[42] \"Parto\"                                    \n",
       "[43] \"Calo_ponderale\"                           \n",
       "[44] \"Aumento_peso\"                             \n",
       "[45] \"Yogurt\"                                   \n",
       "[46] \"Pane\"                                     \n",
       "[47] \"Pasta\"                                    \n",
       "[48] \"Latticini\"                                \n",
       "[49] \"Frutta.Verdua\"                            \n",
       "[50] \"Carne\"                                    \n",
       "[51] \"Pesce\"                                    \n",
       "[52] \"Cereali\"                                  \n",
       "[53] \"Legumi\"                                   \n",
       "[54] \"Caffè\"                                    \n",
       "[55] \"Alcool\"                                   \n",
       "[56] \"Pizza\"                                    \n",
       "[57] \"Fumo_al_momento_del_prelievo\"             \n",
       "[58] \"Esercizio_Fisico\"                         \n",
       "[59] \"Comorbidità_degne_di_nota\"                \n",
       "[60] \"Comorbidità_prichiatriche\"                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(sample_data(phyG))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as.data.frame(tax_table(phy))[as.data.frame(tax_table(phy))$Order == \"Clostridiales\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:1:3: unexpected string constant\n16: \n17: clost <- mat[c(which(grepl('\n      ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:1:3: unexpected string constant\n16: \n17: clost <- mat[c(which(grepl('\n      ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "'''#agglomerazione per genere\n",
    "phyG <-tax_glom(phy,\"Genus\")\n",
    "\n",
    "mat <- t(cbind(as(otu_table(phyG), \"matrix\"), as(tax_table(phyG)[,\"Genus\"], \"matrix\")))\n",
    "mat <- as.data.frame(mat)\n",
    "ID <- append(c(sample_data(phyG)$ID_original),\"NA\")\n",
    "IBD <- append(c(sample_data(phyG)$IBD),\"NA\")\n",
    "mat$ID <- ID\n",
    "mat$IBD <- IBD\n",
    "\n",
    "mat$IBD <- revalue(mat$IBD, c(\"2\"=\"1\", \"1\"=\"0\"))\n",
    "\n",
    "write.csv(t(mat),\"reads4Genus.csv\")\n",
    "\n",
    "mat <- as.data.frame(t(mat))\n",
    "\n",
    "clost <- mat[c(which(grepl('Clostridium',mat$Genus,fixed=FALSE)),279,280),]\n",
    "clost_ <- mat[c(which(grepl('clostridium',mat$Genus,fixed=FALSE)),279,280),]\n",
    "\n",
    "clostridium <- rbind(clost[1:5,],clost_)\n",
    "\n",
    "clostridium <- as.data.frame(t(clostridium))\n",
    "clostridium$IBD <- revalue(clostridium$IBD, c(\"2\"=\"1\", \"1\"=\"0\"))\n",
    "\n",
    "write.csv(clostridium,\"reads4Genus_clostridium.csv\")\n",
    "\n",
    "sel <- c(\"21\",\"54\" ,\"59\",\"63\",\"66\",\"69\",\"72\",\"83\",\"87\",\"91\",\"97\",\"98\",\"101\",\"102\",\"107\",\"1\",\"9\",\"9\",\"15\",\"19\",\"22\",\"24\",\"25\",\"37\",\"38\",\"42\",\"51\",\"52\",\"63\",\"62\",\"65\",\"68\")\n",
    "\n",
    "antony <- clostridium %>% filter( ID %in% sel)\n",
    "traspose <- as.data.frame(t(clostridium))\n",
    "antony <- as.data.frame(t(antony))\n",
    "antony$Genus <- traspose$Genus\n",
    "\n",
    "#write.csv(antony,\"reads4Genus_clostridium_antonio.csv\") '''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalization\n",
    "\n",
    "The large variability of the total counts per sample prevents meaningful comparisons of raw abundances between individuals, therefore it is needed to normalize data before any statistical analysis.\n",
    "There are different method to normalize:\n",
    "\n",
    "<ul>\n",
    "    <li> Computation of relative abundances by dividing the raw abundance by the total number of counts per sample </li>\n",
    "    <li> Rarefaction, which consists on subsampling the same number of reads for each sample, so that every samples have the same number of total counts ( not raccomanded beacause it leads to a loss of information. </li>\n",
    "    <li> Deseq2 normalization tecnique, which is the most realiable. </li>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rarefaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### normalizzazione con rarefazione\n",
    "\n",
    "phy_rar <- rarefy_even_depth(phyS, sample.size=6000, rngseed=1)\n",
    "##\n",
    "G1 <- testPermanovaPCoA(phy_rar,DistancePhyl=\"bray\",Variable=\"IBD\",Permutations=9999)\n",
    "G2 <- testPermanovaPCoA(phy_rar,DistancePhyl=\"unifrac\",Variable=\"IBD\",Permutations=9999)\n",
    "G3 <- testPermanovaPCoA(phy_rar,DistancePhyl=\"wunifrac\",Variable=\"IBD\",Permutations=9999)\n",
    "G4 <- testPermanovaPCoA(phy_rar,DistancePhyl=\"canberra\",Variable=\"IBD\",Permutations=9999)\n",
    "##\n",
    "library(\"ggpubr\")\n",
    "P <- ggarrange(G1,G2,G3,G4,ncol=2,nrow=2,labels=c(\"A\",\"B\",\"C\",\"D\"))\n",
    "##\n",
    "#Save.graph(P, OutDir=\"Immagini\" , Name=\"PCoA_Silva_Rarefazione_IBD\" , Name_prefix=\"001\")\n",
    "P"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deseq2\n",
    "\n",
    "High-throughput microbiome datasets generated by the 16S ribosome RNA (rRNA) gene sequencing or shotgun metagenomic sequencing have some properties that require tailored analytic tools; these include count compositional structure, <b>varied total sequence reads across samples</b>, <b>over-dispersion</b> and <b>zero-inflation</b>. \n",
    "\n",
    "One way to account for varing total reads is normalization, i.e., conversion of the sequence counts to the relative abundance (or proportion) using the total sum, mean, or median of representative OTUs across all samples. The negative binomial regression, which is a standard statistical method for analyzing over-dispersed count observations, has been recently applied to microbiome data. On the other hand, several zero-inflated models have also been proposed to correct for excess zero counts in microbiome measurements, including zero-inflated Gaussian, lognormal, negative bimomial and beta models.\n",
    "\n",
    "The <b>negative binomial distribution</b> also arises as a continuous mixture of Poisson distributions (i.e. a compound probability distribution) where the mixing distribution of the Poisson rate is a gamma distribution. Read counts for a sample in theory follow a Binomial(n,p) distribution, where n is the total number of reads and p is the probability of a read mapping to a specific gene. However, the binomial distribution is computationally inconvenient, since it involves computing factorials, and with millions of reads in each sample, the Poisson distribution (with lambda = n*p) is an excellent approximation to the Binomial while being far more mathematically tractable. So the Poisson noise quantifies the counting uncertainty, while the gamma distribution quantifies the variation in gene expression between replicates. The mixture of the two yields the negative binomial.\n",
    "The negative binomial is a \n",
    "\n",
    "DESeq2’s median of ratios consists in dividing counts by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene (OTU). This is performed by dividing each raw count value in a given sample by that sample’s normalization factor to generate normalized count values. This is performed for all count values (every gene in every sample). Log-ratio approach has been introduced by Aitchison and laid the foundations of Compositional Data Analysis (CoDA). Mathematically, the assertion that the relevant information is contained in the ratios between the components implies that two proportional compositions are equally informative and this induces equivalence classes of vectors carrying the same information. In microbiome analysis, for example, both the raw counts and their transformation into relative abundances or proportions belong to the same equivalence class and they carry the same relative information.\n",
    "\n",
    "<img src=\"Istantanea_2020-02-18_10-16-12.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#otu_table(phy) <- otu_table(phy)\n",
    "\n",
    "#trasforma l'oggetto phyloseq in un oggetto deseq2\n",
    "Q <- phyloseq_to_deseq2(phyF, ~ IBD)\n",
    "\n",
    "#calcolo della media geometrica\n",
    "gm_mean <- function(x, na.rm=TRUE){\n",
    "    exp(sum(log(x[x>0]),na.rm=na.rm) / length(x) )\n",
    "}\n",
    "\n",
    "# normalizza usando DESeq2, applicando la media geometrica ai dati\n",
    "geoMeans = apply(counts(Q),1,gm_mean)\n",
    "\n",
    "#applica un fattore di correzione\n",
    "# le OTU con un numero di reads molto alto possono portare a sottostimare \n",
    "#l'abbondanza di altre OTU con un numero di reads più basso\n",
    "D <- estimateSizeFactors(Q, geoMeans = geoMeans)\n",
    "\n",
    "normalized_count <- as.data.frame(round(counts(D,normalized=TRUE)))\n",
    "\n",
    "otu_table(phyF) <- otu_table(normalized_count, taxa_are_rows=TRUE)\n",
    "\n",
    "# salva in un dataframe per poterli poi ricaricare\n",
    "#write.table(normalized_count,\"DADA2_output/deseq2_otu_normalized.txt\",sep=\"\\t\", quote=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Diversity analisys\n",
    "\n",
    "The diversity of the microbiome is an important indicator of the good or bad conditions of the ecosystem, with larger microbiome diversity being usually associated to better health status. Microbiome diversity can be assessed through multiple ecological indices that can be divided into two kind of measures:\n",
    "<ol>\n",
    "    <li><b>Alpha diversity</b> which measures the variability of species within a sample. </li>\n",
    "    <li><b>Beta diversity</b> which accounts for the differences in composition between samples.</li>\n",
    "</ol>\n",
    "The R package <b>vegan</b> provides a large set of diversity measures.\n",
    "\n",
    "### Alpha diversity\n",
    "\n",
    "Alpha diversity within sample diversity The most important measure of alpha diversity is <b>richness</b>, defined as the number of different species present in an environment. Richness is estimated by the observed richness, the number of different species observed in the sample. The observed richness tends to underestimate the real richness in the environment, where the less frequent species are likely to be undetected. There are different indices that adjust for this and try to estimate the hidden part that has not been detected. One of the most extended richness measure is <b>Chao1 index</b>. \n",
    "\n",
    "Another important indicator of alpha diversity is <b>evenness</b>, which measures the homogeneity in abundance of the different species in a sample. A commonly used measure of evenness is the <b>Shannon index</b>.\n",
    "\n",
    "### Beta diversity\n",
    "\n",
    "Beta diversity measures the differences in microbiome composition between samples. There is a wide range of ecological distances or dissimilarities for measuring how close are two microbial compositions. The most commonly used are <b>Bray-Curtis</b>, <b>UniFrac</b> and <b>weighted UniFrac</b> distances. We also define the Aitchison distance which is a proper distance for compositional data.\n",
    "<ul>\n",
    "<b>UniFrac</b> family of distances consider the phylogenetic tree that represents the evolutionary relationships among the different taxa. The phylogenetic tree can be obtained from the bioinformatic pipelines, such as mothur and <b>QIIME</b>. For a tree with r branches, let b = (b<sub>1</sub>,…,b<sub>r</sub>) represent the length of the different branches in the phylogenetic tree, and q<sub>1</sub> = (q<sub>11</sub> , … ,q<sub>1r</sub> ), and q<sub>2</sub> = (q<sub>21</sub> ,… ,q<sub>2r</sub> ) the relative abundances associated to each branch for the first and the second sample, respectively.\n",
    "<li><b>Unweighted UniFrac</b> distance measures the relative length of those branches that lead exclusively to species present in only one of the two samples with respect to the total length of all branches in the tree. It only takes into account the presence or absence of the taxa.\n",
    "</li>\n",
    "<li><b>Weighted UniFrac</b> distance instead includes information on the relative abundance of each taxa.\n",
    "</li>\n",
    "</ul>\n",
    "\n",
    "### Ordination \n",
    "\n",
    "The goal of ordination plots is the visualization of beta diversity for identification of possible data structures. The multidimensional data is represented into a reduced number of orthogonal axes while keeping the main trends of the data and preserving the distances among samples as much as possible. Most commonly used ordination methods for microbiome data are <b>principal coordinates analysis (PCoA)</b>, also known as multidimensional scaling, and non-metric multidimensional scaling (NMDS). \n",
    "\n",
    "PCoA an extension of Principal Components Analysis (PCA). Given a distance or dissimilarity matrix, D, PCoA performs eigenvalue decomposition of D<sub>c</sub>'D<sub>c</sub> where D<sub>c</sub> is the centered distance matrix. When D is the Euclidean distance, PCoA results exactly the same as PCA. Care must be taken with PCoA if the selected distance is not metric, because some eigenvalues may be negative and then, the graphical representation will not perform properly. In order to avoid this problem NMDS is more commonly used. Also based on a distance matrix D, NMDS maximizes the rank-based correlation between the original distances and the distances between samples in the new reduced ordination space. The procedure starts with a random configuration and the optimal representation is obtained following an iterative procedure that at each steps improves the rank correlation. \n",
    "\n",
    "Ordination plots can be obtained with the R and Bioconductor packages vegan and phyloseq, among others. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Microbiome Statistical Inference\n",
    "\n",
    "### Multivariate differential abundance testing \n",
    "\n",
    "Multivariate differential abundance testing refers to a <b>global test of differences in microbial composition</b> between two or more groups of samples. We can distinguish between distance-based or model-based approaches:\n",
    "\n",
    "<ol>\n",
    "<li> The most widely used distance-based method for multivariate community analysis is PERMANOVA(Permutational Multivariate Analysis of Variance Using Distance Matrices). The null hypothesis of no differences in composition among groups is formulated by the condition that the different groups of samples have the same center of masses.\n",
    "    \n",
    "Implemented in the function <b>\"adonis\"</b> of the vegan R package, it consists of a multivariate ANOVA based on dissimilarities. The variability within groups is compared against the variability between groups with the usual ANOVA F statistic, but partition of sums-of-squares is applied directly to dissimilarities. \n",
    "Significance is evaluated through permutations to generate a distribution of the pseudo F statistic under the null hypotesis. A related and popular distance-based approach is the analysis of similarities, implemented in the function “anosim” of the vegan R package.\n",
    "</li>\n",
    "<p>\n",
    "<li>An interesting model-based approach for multivariate microbiome analysis is Kernel machine regression (KMR), that extends PERMANOVA to a regression framework. KMR is a semi-parametric regression model that includes a nonparametric component. The model can be expressed as a semiparametric linear regression model when the response variable is continuous or as a semiparametric logistic regression model for a dichotomous response variable.\n",
    "The non-parametric component is related to a Kernel matrix that is a transformation of the distance matrix D of pairwise distances between individuals. </li>\n",
    "</ol>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PERMANOVA\n",
    "\n",
    "Permutational multivariate analysis of variance (PERMANOVA), is a <u>non-parametric multivariate statistical test</u>. PERMANOVA is used to compare groups of objects and test the null hypothesis that the centroids and dispersion of the groups as defined by measure space are equivalent for all groups. A rejection of the null hypothesis means that either the centroid and/or the spread of the objects is different between the groups. Hence the test is based on the prior calculation of the distance between any two objects included to the experiment. PERMANOVA shares some resemblance to ANOVA where they both measure the sum-of-squares within and between group and make use of F test to compare within-group to between-group variance. However, while ANOVA bases the significance of the result on assumption of normality, PERMANOVA draws tests for significance by comparing the actual F test result to that gained from random permutations of the objects between the groups. Moreover, while PERMANOVA tests for similarity based on a chosen distance measure, ANOVA tests for similarity of the group averages.\n",
    "\n",
    "In the simple case of a single factor with p groups and n objects in each group,the total sum-of-squares is determined as:\n",
    "\n",
    "<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/cfac536edd1016aa559d1ccf78e48679a67c2e04\">\n",
    "\n",
    "where N is the total number of objects, and d<sub>ij</sub> is the squared distance between objects i and j.\n",
    "\n",
    "Similarly, the within groups sum-of-squares is determined:\n",
    "\n",
    "<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/c2d26a665b1f60b98299b888b56e46d4bf3dbfe9\">\n",
    "\n",
    "where  ε<sub>ij</sub> takes the value of 1 if observation i and observation j are in the same group, otherwise it takes the value of zero. Then, the between groups sum-of-squares <i>SS<sub>A</sub></i> can be calculated as the difference between the overall and the within groups sum-of-squares:\n",
    "\n",
    "<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/da0a2898b61f40fb4bf77787edcae82c119c35f5\">\n",
    "\n",
    "Finally, a <b>pseudo F-statistic</b> is calculated:\n",
    "\n",
    "<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/2399314b3c10e2b1dc24bfe4beda2d665d1f8e14\">\n",
    "where p is the number of groups.\n",
    "\n",
    "Finally, the PERMANOVA procedure draws significance for the actual F statistic by performing multiple permutations of the data. In each such the items are shuffled between groups. For each such permutation of the data the permutation F statistic is calculated. \n",
    "\n",
    "The <b>p-value</b> is then calculated by:\n",
    "\n",
    "<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/e48ec1eb507b40fd2804f7cd1112b7c7d767f0c2\">\n",
    "\n",
    "Where <i>F</i> is the F statistic obtained from the original data and <i>F<sup>p</sup></i> is a permutation F statistic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#normalized_count <- read.table(\"DADA2_output/deseq2_otu_normalized.txt\",sep=\"\\t\")\n",
    "\n",
    "colnames(normalized_count) <- gsub(\"\\\\.\",\"-\",colnames(normalized_count))\n",
    "otu_table(phyF) <- otu_table(normalized_count, taxa_are_rows=TRUE)\n",
    "\n",
    "\n",
    "G1 <- testPermanovaPCoA(phyF,DistancePhyl=\"bray\",method=\"NMDS\",Variable=\"IBD\",Permutations=9999)\n",
    "G2 <- testPermanovaPCoA(phyF,DistancePhyl=\"unifrac\",method=\"NMDS\",Variable=\"IBD\",Permutations=9999)\n",
    "G3 <- testPermanovaPCoA(phyF,DistancePhyl=\"wunifrac\",method=\"NMDS\",Variable=\"IBD\",Permutations=9999)\n",
    "G4 <- testPermanovaPCoA(phyF,DistancePhyl=\"canberra\",method=\"NMDS\",Variable=\"IBD\",Permutations=9999)\n",
    "G5 <- testPermanovaPCoA(phyF,DistancePhyl=\"euclidean\",method=\"NMDS\",Variable=\"IBD\",Permutations=9999)\n",
    "\n",
    "library(\"ggpubr\")\n",
    "P <- ggarrange(G1,G2,G3,G4,ncol=2,nrow=2,labels=c(\"A\",\"B\",\"C\",\"D\"))\n",
    "\n",
    "\n",
    "Save.graph(P, OutDir=\"Immagini\" , Name=\"NMDS_Silva_NormDESeq2_IBD_Family\" , Name_prefix=\"001\")\n",
    "\n",
    "P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#normalized_count <- read.table(\"DADA2_output/deseq2_otu_normalized.txt\",sep=\"\\t\")\n",
    "\n",
    "colnames(normalized_count) <- gsub(\"\\\\.\",\"-\",colnames(normalized_count))\n",
    "otu_table(phyF) <- otu_table(normalized_count, taxa_are_rows=TRUE)\n",
    "\n",
    "\n",
    "G1 <- testPermanovaPCoA(phyF,DistancePhyl=\"bray\",method=\"PCoA\",Variable=\"IBD\",Permutations=9999)\n",
    "G2 <- testPermanovaPCoA(phyF,DistancePhyl=\"unifrac\",method=\"PCoA\",Variable=\"IBD\",Permutations=9999)\n",
    "G3 <- testPermanovaPCoA(phyF,DistancePhyl=\"wunifrac\",method=\"PCoA\",Variable=\"IBD\",Permutations=9999)\n",
    "G4 <- testPermanovaPCoA(phyF,DistancePhyl=\"canberra\",method=\"PCoA\",Variable=\"IBD\",Permutations=9999)\n",
    "G5 <- testPermanovaPCoA(phyF,DistancePhyl=\"euclidean\",method=\"PCoA\",Variable=\"IBD\",Permutations=9999)\n",
    "\n",
    "library(\"ggpubr\")\n",
    "P <- ggarrange(G1,G2,G3,G4,ncol=2,nrow=2,labels=c(\"A\",\"B\",\"C\",\"D\"))\n",
    "\n",
    "\n",
    "Save.graph(P, OutDir=\"Immagini\" , Name=\"PCoA_Silva_NormDESeq2_IBD_Family\" , Name_prefix=\"001\")\n",
    "\n",
    "P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "[1] \"D_bray  ~  IBD+Genere+Eta+Yogurt+Pasta+Pane+Latticini+FruttaVerdua+Carne+Cereali+Legumi+Alcool+Esercizio_Fisico+Fumo_al_momento_del_prelievo\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in qr.fitted(qrhs, G): 'qr' and 'y' must have the same number of rows\n",
     "output_type": "error",
     "traceback": [
      "Error in qr.fitted(qrhs, G): 'qr' and 'y' must have the same number of rows\nTraceback:\n",
      "1. PERMANOVA_stepwise_backward(metadata, distance_matrix, parameters = Formula)",
      "2. adonis2(as.formula(Formula), data = metadata, permutations = 9999, \n .     by = \"margin\")",
      "3. adonis0(formula, data = data, method = method)",
      "4. qr.fitted(qrhs, G)",
      "5. stop(\"'qr' and 'y' must have the same number of rows\")"
     ]
    }
   ],
   "source": [
    "metadata <- as(sample_data(phyF),\"data.frame\")\n",
    "\n",
    "\n",
    "D_bray <- distance(phyF,method=\"bray\")\n",
    "D_Uuni <- distance(phyF,method=\"unifrac\")\n",
    "D_Wuni <- distance(phyF,method=\"wunifrac\")\n",
    "D_C <- distance(phyF,method=\"canberra\")\n",
    "\n",
    "distance_matrix <- \"D_bray\"\n",
    "Formula <- \"IBD+Genere+Eta+Yogurt+Pasta+Pane+Latticini+FruttaVerdua+Carne+Cereali+Legumi+Alcool+Esercizio_Fisico+Fumo_al_momento_del_prelievo\"\n",
    "PERMANOVA_stepwise_backward(metadata,distance_matrix,parameters=Formula)\n",
    "\n",
    "distance_matrix <- \"D_Uuni\"\n",
    "Formula <- \"IBD+Genere+Eta+Yogurt+Pasta+Pane+Latticini+FruttaVerdua+Carne+Cereali+Legumi+Alcool+Esercizio_Fisico+Fumo_al_momento_del_prelievo\"\n",
    "PERMANOVA_stepwise_backward(metadata,distance_matrix,parameters=Formula)\n",
    "\n",
    "distance_matrix <- \"D_Wuni\"\n",
    "Formula <- \"IBD+Genere+Eta+Yogurt+Pasta+Pane+Latticini+FruttaVerdua+Carne+Cereali+Legumi+Alcool+Esercizio_Fisico+Fumo_al_momento_del_prelievo\"\n",
    "PERMANOVA_stepwise_backward(metadata,distance_matrix,parameters=Formula)\n",
    "\n",
    "distance_matrix <- \"D_C\"\n",
    "Formula <- \"IBD+Genere+Eta+Yogurt+Pasta+Pane+Latticini+FruttaVerdua+Carne+Cereali+Legumi+Alcool+Esercizio_Fisico+Fumo_al_momento_del_prelievo\"\n",
    "PERMANOVA_stepwise_backward(metadata,distance_matrix,parameters=Formula)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Univariate differential abundance testing \n",
    "\n",
    "When significant global differences in microbiome composition are detected between groups of samples, a natural question arises: which particular taxa are responsible of that global difference? \n",
    "\n",
    "A common strategy to answer this question is to test every taxa separately for association with the response variable. When the response variable is dichotomous this is known as <b>univariate differential abundance testing</b>. Below are described both, classical and CoDA approaches for univariate differential abundance testing. However, we advise that classical univariate approaches are notably affected by the compositional structure of microbiome data and their results, with large false discovery rates, might be questioned.\n",
    "\n",
    "Nonparametric tests, like the <b>Wilcoxon rank-sum test</b> or the <b>Kruskal-Wallis test</b>, can be applied. However, more powerful parametric approaches are available, such as the Bioconductor packages edgeR and DESeq2, initially proposed for transcriptomics analysis (RNA-Seq data). \n",
    "\n",
    "<u>Both fit a generalized linear model and assume that read counts follow a Negative Binomial distribution</u>. The NB distribution extends the Poisson distribution by allowing the variance to be different from the mean. edgeR and DESeq2 mainly differ in the way they normalize the data.<u> DESeq2 uses size factors that account for differences in sequencing depth between samples and shrinkage for large variances correction</u>. edgeR can be implemented with different normalization methods but the most recommended is TMM. Two CoDA methods that explicitly accounts for the compositional nature of microbiome data are <b>ANCOM</b> and <b>ALDEx2</b>. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#level <- \"Genus\"\n",
    "level <- \"Family\"\n",
    "#phyLevel <-tax_glom(phy,level)\n",
    "#phyLevel <- phyG\n",
    "phyLevel <- phyF\n",
    "\n",
    "Q <- phyloseq_to_deseq2(phyLevel, ~ IBD)\n",
    "\n",
    "#calcolo della media geometrica\n",
    "gm_mean <- function(x, na.rm=TRUE){\n",
    "    exp(sum(log(x[x>0]),na.rm=na.rm) / length(x) )\n",
    "}\n",
    "\n",
    "# normalizza usando DESeq2, applicando la media geometrica ai dati\n",
    "geoMeans = apply(counts(Q),1,gm_mean)\n",
    "\n",
    "#applica un fattore di correzione\n",
    "# le OTU con un numero di reads molto alto possono portare a sottostimare \n",
    "#l'abbondanza di altre OTU con un numero di reads più basso\n",
    "D <- estimateSizeFactors(Q, geoMeans = geoMeans)\n",
    "\n",
    "normalized_count <- as.data.frame(round(counts(D,normalized=TRUE)))\n",
    "\n",
    "otu_table(phyLevel) <- otu_table(normalized_count, taxa_are_rows=TRUE)\n",
    "\n",
    "\n",
    "dds <- DESeq(D)\n",
    "res <- results(dds)\n",
    "\n",
    "res <- as.data.frame(res)\n",
    "\n",
    "res <- res[order(res$padj),]\n",
    "\n",
    "alpha = 0.05\n",
    "sigtab = res[which(res$padj < alpha), ]\n",
    "#sigtab = res[which(res$log2FoldChange < -2), ]\n",
    "sigtab = cbind(as(sigtab, \"data.frame\"), as(tax_table(phyLevel)[rownames(sigtab), ], \"matrix\"))\n",
    "sigtab <- sigtab[order(sigtab$padj),]\n",
    "table <-sigtab[,c(\"log2FoldChange\",\"padj\",level)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table <- table[table$Family != \"uncultured\" , ]\n",
    "table <- table[table$Family !=\"gut metagenome\",]\n",
    "\n",
    "row.names(table) <- table$Family\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barplotting_variables<- function(genus,tassonomia,level,phy_percent){\n",
    "    for (r in 1:nrow(tassonomia)){\n",
    "\n",
    "\ttaxa <- as.character(tassonomia[r,]$Family)\n",
    "\t# rimuove il punto che da problemi quando si salva il grafico\n",
    "\ttaxa <- gsub('\\\\.', '', taxa)\n",
    "\ttaxa <- gsub(':', '', taxa)\n",
    "    #print(type(taxa))\n",
    "\tID_taxa <- rownames(tassonomia[r,])\n",
    "\tcat(\"Analisi della tassonomia: \",taxa,\" con ID: \",ID_taxa,\"\\n\")\n",
    "\t\n",
    "\t# analisi per valutare la concentrazione tra IBD = 0 vs IBD = 1\n",
    "\tphy_percentuali_reduced <- take_taxa(phy_percent, ID_taxa)\n",
    "\t#new_order <- c(\"21\",\"54\",\"59\",\"63\",\"66\",\"69\",\"72\",\"83\",\"87\",\"91\",\"97\",\"98\",\"101\",\"102\",\"107\",\"46\",\"70\",\"1\",\"9\",\"19\",\"22\",\"24\",\"25\",\"37\",\"38\",\"42\",\"51\",\"52\",\"53\",\"62\",\"65\",\"68\",\"4\",\"8\",\"10\",\"12\",\"17\",\"18\",\"23\",\"33\",\"35\",\"40\",\"41\",\"44\",\"55\",\"57\",\"58\",\"64\",\"71\",\"77\",\"80\",\"81\",\"85\",\"88\",\"92\",\"96\",\"99\",\"100\",\"103\",\"104\",\"109\",\"110\",\"112\")\n",
    "\tmy_colors<-c(\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"violetred4\",\"violetred4\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\")\n",
    "\tmetadata <- sample_data(phy_percentuali_reduced)\n",
    "\tmetadata$ID_original <- factor(metadata$ID_original) #, levels = new_order )\n",
    "\tsample_data(phy_percentuali_reduced) <- metadata\n",
    "\tP <- plot_bar(phy_percentuali_reduced,  x=\"ID_original\", fill=\"IBD\") + facet_grid(~IBD,scales=\"free_x\")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +facet_wrap(~IBD,scales=\"free_x\") +ggtitle(taxa)\n",
    "\t#P <- plot_bar(phy_percentuali_reduced,  x=\"ID_original\", fill=\"IBD\", facet_grid=~IBD) + facet_grid(scales=\"free_x\") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +facet_wrap(~IBD) +ggtitle(genus)\n",
    "\t#P <- plot_bar(phy_percentuali_reduced,  x=\"ID_original\", fill=\"IBD\", facet_grid=~IBD) + facet_grid(scales=\"free_x\") \n",
    "\tSave.graph( Graph=P, OutDir=\"Immagini\", Name=taxa, Name_prefix=paste(taxa,\"IBD_0_1_\") )\n",
    "\t\n",
    "\t# adesso possiamo provare a fare a stampare in funzione delle varie scale montreal, riducendo ai soli pazienti\n",
    "\tphy_percentuali_reduced_patients <- subset_samples( phy_percentuali_reduced, IBD == \"1\" )\n",
    "\tmy_colors_due<-c(\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"darkgreen\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\",\"violetred4\")\n",
    "\n",
    "\tP1 <- plot_bar(phy_percentuali_reduced_patients, x=\"ID_original\", fill=\"Montreal_MDC_localizzazione\", facet_grid=~Montreal_MDC_localizzazione) + facet_grid(scales=\"free_x\") +\n",
    "\t\ttheme(axis.text.x = element_text(angle = 45, hjust = 1))+ ggtitle(taxa) #+ facet_wrap(~Montreal_MDC_localizzazione, scales=\"free_x\")\n",
    "        \n",
    "\tP2 <- plot_bar(phy_percentuali_reduced_patients, x=\"ID_original\", fill=\"Montreal_MDC_Eta\", facet_grid=~Montreal_MDC_Eta) + facet_grid(scales=\"free_x\") +\n",
    "\t\ttheme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(taxa) #+facet_wrap(~Montreal_MDC_Eta,scales=\"free_x\")\n",
    "        \n",
    "\tP3 <- plot_bar(phy_percentuali_reduced_patients, x=\"ID_original\", fill=\"Montreal_MDC_Fenotipo\", facet_grid=~Montreal_MDC_Fenotipo) + facet_grid(scales=\"free_x\") +\n",
    "\t\ttheme(axis.text.x = element_text(angle = 45, hjust = 1))+ ggtitle(taxa)#+facet_wrap(~Montreal_MDC_Fenotipo, scales=\"free_x\")\n",
    "        \n",
    "\t\n",
    "\tSave.graph( Graph=P1, OutDir=\"Immagini\", Name=taxa, Name_prefix=paste(genus,\"IBD_Montreal_Loc_\"))\n",
    "\tSave.graph( Graph=P2, OutDir=\"Immagini\", Name=taxa, Name_prefix=paste(genus,\"IBD_Montreal_Eta_\"))\n",
    "\tSave.graph( Graph=P3, OutDir=\"Immagini\", Name=taxa, Name_prefix=paste(genus,\"IBD_Montreal_Fen_\"))\n",
    "\t\n",
    "\t\n",
    "\tcat(\"\\n\")\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyG_percentuali"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "generi <- as.vector(table$Family)\n",
    "#generi <- c(\"\\\\[Eubacterium\\\\] ruminantium group\",\"\\\\[Eubacterium\\\\] xylanophilum group\",\"\\\\[Eubacterium\\\\] ventriosum group\")\n",
    "for(genus in generi){\n",
    "    print(genus)\n",
    "    if (genus != \"uncultured\" && genus != \"gut metagenome\"){\n",
    "        righe <- grepl(genus,tassonomiaG$Family,fixed=FALSE,ignore.case=TRUE)\n",
    "        tassonomia_genus <- tassonomiaG[ righe , ]\n",
    "        barplotting_variables(genus, tassonomia_genus,level,phyG_percentuali)\n",
    "        } \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "generi <- as.vector(table$Genus)\n",
    "df <- data.frame(Genus=character(), Pvalue=numeric(), Chisquared = numeric(), df=numeric())\n",
    "for(genus in generi){\n",
    "    if (genus== \"Escherichia-Shigella\" | genus== \"Enterococcus\" | genus ==\"Lactobacillus\"){\n",
    "        print(genus)\n",
    "        righe <- grepl(genus,tassonomiaG$Genus,fixed=FALSE,ignore.case=TRUE)\n",
    "        tassonomia_genus <- tassonomiaG[ righe , ]\n",
    "        for (r in 1:nrow(tassonomia_genus)){\n",
    "            taxa <- as.character(tassonomia_genus[r,]$level)\n",
    "            taxa <- gsub('\\\\.', '', taxa)\n",
    "            taxa <- gsub(':', '', taxa)\n",
    "            ID_taxa <- rownames(tassonomia_genus[r,])\n",
    "            cat(\"Analisi della tassonomia: \",taxa,\" con ID: \",ID_taxa,\"\\n\")\n",
    "        \n",
    "            # analisi per valutare la concentrazione tra IBD = 0 vs IBD = 1\n",
    "            phy_percentuali_reduced <- take_taxa(phyG_percentuali, ID_taxa)\n",
    "            phy_percentuali_reduced <- subset_samples( phy_percentuali_reduced , Montreal_MDC_localizzazione != \"\" )\n",
    "            #test_differential_abundance_Kruskal(phy_percentuali_reduced, Montreal_MDC_localizzazione, compare = NULL,\n",
    "            #block = , p.adjust.method = \"fdr\")\n",
    "            phy_df <- as.data.frame(t(as.data.frame(otu_table(phy_percentuali_reduced))))\n",
    "        \n",
    "            phy_df$Montreal_MDC_localizzazione <- sample_data(phy_percentuali_reduced)$Montreal_MDC_localizzazione\n",
    "        \n",
    "            colnames(phy_df) <- c(\"Genus\",\"Montreal_MDC_localizzazione\")\n",
    "        \n",
    "            res <- kruskal.test(Genus ~ Montreal_MDC_localizzazione, data = phy_df )\n",
    "            row <- data.frame(c(genus),c(res$p.value),c(res$statistic),c(res$parameter))\n",
    "            colnames(row) <- c(\"Genus\",\"Pvalue\",\"Chisquared\",\"df\")\n",
    "            row.names(row) <- c()\n",
    "            df <- rbind(df,row)\n",
    "            \n",
    "            }\n",
    "        \n",
    "        } \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(df,\"Montreal_MDC_localizzazione_kruskal.test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d <- as.data.frame(otu_table(phy))\n",
    "\n",
    "d <- t(d[rownames(table),])\n",
    "\n",
    "tax <- tax_table(phy)[rownames(table),level]\n",
    "\n",
    "merged <- merge(t(d), tax, by=\"row.names\",all.x=TRUE )\n",
    "\n",
    "d1 <- sample_data(phy)[,\"IBD\"]\n",
    "\n",
    "d <-merged\n",
    "\n",
    "d$Genus <- NULL\n",
    "\n",
    "row.names(d)<-merged$Genus\n",
    "\n",
    "d$Row.names <- NULL\n",
    "\n",
    "d <- t(d)\n",
    "\n",
    "d <- as.data.frame(merge(d,d1,by=\"row.names\",all.x=TRUE))\n",
    "\n",
    "temp <- d\n",
    "\n",
    "d$Row.names <- NULL\n",
    "\n",
    "row.names(d) <- temp$Row.names\n",
    "\n",
    "#write.csv(df,\"few_analysis.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#range01 <- function(x){(x-min(x))/(max(x)-min(x))}\n",
    "\n",
    "#d$value <- range01(d$value)\n",
    "\n",
    "d <- melt(data = d, Genus = \"Genus\", measure.vars = colnames(d)[1:9])\n",
    "\n",
    "families <- t(tax_table(phy)[,\"Genus\"])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "png(\"few_analysis.png\")\n",
    "p <-  ggplot(d, aes(x=variable, y=value, fill=IBD)) + \n",
    "    geom_boxplot()+facet_wrap(vars(variable), scales=\"free\")\n",
    "print(p)\n",
    "dev.off()\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#seleziono solo i pazienti che hanno il clostridium\n",
    "phyG_innocuum <- subset_samples(phyF, ID_original == \"1\" | ID_original == \"22\" | ID_original == \"37\" | ID_original == \"38\" | ID_original == \"57\" |\n",
    "    ID_original == \"58\" | ID_original == \"77\" | ID_original == \"80\" | ID_original == \"81\" | ID_original == \"85\" | ID_original == \"103\" |\n",
    "    ID_original == \"109\" | ID_original == \"110\" | ID_original == \"8\" | ID_original == \"10\" | ID_original == \"12\" | ID_original == \"17\" |\n",
    "    ID_original == \"33\" | ID_original == \"35\")\n",
    "\n",
    "df_innocuum <- as.data.frame(sample_data(phyG_innocuum)$Montreal)\n",
    "values <- c(0.045932135270138376, 0.23414329569696654, 0.7023875624652971, 0.05609835912299565, 0.2563267963219139, 0.017537321612556722, 0.1222759096736022, 0.04022028339830456, 0.3533084073680239, 0.05528193788320434, 0.25521259690855685, 0.03292045594831488, 0.1080580426057425, 0.07802678919762453, 0.14455862480751705, 0.1788508830762352, 0.5090977656264731, 0.032783953744821626, 0.14086367037901132)\n",
    "colnames(df_innocuum)<-\"Montreal\"\n",
    "row.names(df_innocuum)<- sample_data(phyG_innocuum)$ID_original\n",
    "df_innocuum$percent_reads<- values\n",
    "df_innocuum <- df_innocuum[order(df_innocuum$percent_reads),]\n",
    "\n",
    "#seleziono i pazienti che non hanno il clostridium\n",
    "phyG_NO_innocuum <- subset_samples(phyF, ID_original == '9'| ID_original == '19'| ID_original == '24'| ID_original == '25'| ID_original == '42'| ID_original == '51'| ID_original == '52'| ID_original == '62'| ID_original == '65'| ID_original == '68'| ID_original == '40'| ID_original == '41'| ID_original == '44'| ID_original == '53'| ID_original == '55'| ID_original == '64'| ID_original == '71'| ID_original == '88'| ID_original == '92'| ID_original == '96'| ID_original == '99'| ID_original == '100'| ID_original == '104'| ID_original == '112'| ID_original == '4'| ID_original == '18'| ID_original == '23')\n",
    "\n",
    "df_NO_innocuum <- as.data.frame(sample_data(phyG_NO_innocuum)$Montreal)\n",
    "colnames(df_NO_innocuum)<-\"Montreal\"\n",
    "row.names(df_NO_innocuum)<- sample_data(phyG_NO_innocuum)$ID_original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"ggplot2\")\n",
    "theme_set(theme_bw())\n",
    "scale_fill_discrete <- function(palname = \"Set1\", ...) {\n",
    "    scale_fill_brewer(palette = palname, ...)\n",
    "}\n",
    "# Phylum order\n",
    "x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))\n",
    "x = sort(x, TRUE)\n",
    "sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))\n",
    "ggplot(sigtab, aes(x=Family, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + \n",
    "  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genus order\n",
    "x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))\n",
    "x = sort(x, TRUE)\n",
    "sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))\n",
    "ggplot(sigtab, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + \n",
    "  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "?merge"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
